### ISSUE TITLE
sd_journal_query_unique() is not very unique

### ISSUE THREAD

**Issue by ente76** [systemd/systemd#18075]
**systemd version the issue has been seen with**
systemd 247.2-1

**Used distribution**
archlinux

**Linux kernel version used** (`uname -a`)
5.9.14-arch1-1 #1 SMP PREEMPT Sat, 12 Dec 2020 14:37:12 +0000 x86_64 GNU/Linux

**CPU architecture issue was seen on**
x86_64

**Expected behaviour you didn't see**
Using the example program for sd_journal_query_unique from your own [documentation](https://www.freedesktop.org/software/systemd/man/sd_journal_query_unique.html#Examples)  & adjusting it to field _TRANSPORT (for the reason of having the error more obvious) I expect the following output:
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel

**Unexpected behaviour you saw**
I get the following output:
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=audit
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=kernel
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=kernel
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=kernel
_TRANSPORT=driver
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=kernel
_TRANSPORT=stdout
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=kernel
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=stdout
_TRANSPORT=driver
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=kernel
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=kernel
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=kernel
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=journal
_TRANSPORT=audit
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=driver
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=kernel
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=journal
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=stdout
_TRANSPORT=syslog
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel
_TRANSPORT=syslog
_TRANSPORT=stdout
_TRANSPORT=audit
_TRANSPORT=journal
_TRANSPORT=driver
_TRANSPORT=kernel

**Steps to reproduce the problem**
I prepared some repository with example code:
```
git clone https://gitlab.com/ente76/journald-errors.git
cd journald-errors
gcc -g query_unique -lsystemd -o query_unique
./query_unique
```

If you fail to reproduce, I will prepare some example journal files.

**Comment by ente76** [comment]
I run the test program against another machine. Query unique is returning unique results on that machine.

I made a copy of my local log, opened the copy using "open_directory" and run multiple times, deleting a few files each time. What I observed: each time I deleted a few files, he is repeating the output less often. I am now down to 2 files: he is repeating exactly twice. One could assume query_unique would be repeating unique values for each file. On the other hand: running exactly the same test against a copy of the log of the other machine with hundreds of files, query_unique is returning unique values.

Question: what is the difference between those files? Both machines run arch linux; always up to date; both run with default configuration of journald.
